src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -714,28 +762,27 @@ def _get_private_clauses(self):
                             limit=loop_ancestor,
                             include_self=True)
 
-                        # If a previous value might be needed we mark it as
-                        # firstprivate
-                        if has_been_read or conditional_write:
+                        if has_been_read:
+                            loop_pos = loop_ancestor.loop_body.abs_position
+                            if last_read_position > loop_pos:
+                                # It is read in the same loop_body, so it is
+                                # a race condition
+                                need_sync.add(symbol)
+                            else:
+                                fprivate.add(symbol)
+                        elif conditional_write:
                             fprivate.add(symbol)
```

Comment:
I can see that the logic here is essentially unchanged but I can't remember the details: why does a conditional write (without a read) imply first private? Is it because it may subsequently be read within the loop? Could/should we check for such a read? Please could you add a comment to explain.

Reply:
> Is it because it may subsequently be read within the loop? 

yes, and if the write didn't happen we need its original value. I extended the comment to make this clearer.

> Could/should we check for such a read?

I don't think so because we may want to maintain the original value also after the loop. This may not happen yet because we don't deal well with accesses after the parallel region (this is tracked in #598) but this is independent from this logic (e.g. OpenMP allows being `firstprivate(x) lastprivate(x)` at the same time). Also I put a TODO for write-only accesses before this point that may also make a check here unecessary.

Reply:
OK. Thanks for extending the comments and adding TODOs.